IPIP/0000-gateway-redirects.md

```diff
@@ -0,0 +1,75 @@
+# IPIP 0000: Gateway Redirects
+
+- Start Date: (format: 2022-06-15)
+- Related Issues:
+  - [ipfs/specs/issues/257](https://github.com/ipfs/specs/issues/257)
+  - [ipfs/kubo/pull/8890](https://github.com/ipfs/kubo/pull/8890)
+
+## Summary
+
+Provide support for URL redirects and rewrites for web sites hosted on Subdomain or DNSLink Gateways, thus enabling support for [single-page applications (SPAs)](https://en.wikipedia.org/wiki/Single-page_application).
+
+## Motivation
+
+Web sites often need to redirect from one URL to another, for example, to change the appearance of a URL, to change where content is located without changing the URL, to redirect invalid URLs to a pretty 404 page, or to enable URL rewriting.
+URL rewriting in particular is a critical feature for hosting SPAs, allowing routing logic to be handled by front end code. SPA support is the primary impetus for this RFC.
+
+Currently the only way to handle URL redirects or rewrites is with additional software such as NGINX sitting in front of the Gateway. This software introduces operational complexity and decreases the uniformity of experience when navigating to content hosted on a Gateway, thus decreasing the value proposition of hosting web sites in IPFS.
+
+This RFC proposes the introduction of redirect support for content hosted on Subdomain or DNSLink Gateways, configured via a `_redirects` file residing underneath the root CID of the web site.
+
+## Detailed design
+
+Allow developers to configure redirect support by adding redirect rules to a file named `_redirects` stored underneath the root CID of their web site.
+The format for this file is similar to those of [Netlify](https://docs.netlify.com/routing/redirects/#syntax-for-the-redirects-file) and [Cloudflare Pages](https://developers.cloudflare.com/pages/platform/redirects) but only supporting a subset of their functionality.
+
+The format for the file is `from to [status]`.
+
+- `from` - specifies the path to intercept (can include placeholders and a trailing splat)
+- `to` - specifies the path or URL to redirect to (can include placeholders or splat matched in `from`)
+- `status` - optional HTTP status code (301 if not specified)
+
+Rules in the file are evaluated top to bottom.
+
+For performance reasons this proposal does not include forced redirect support (i.e. redirect rules that are evaluated even if the `from` path exists). In other word, redirect logic will be evaluated if and only if the requested path does not exist.  If the requested path exists, we won't even check for the existence of the `_redirects` file.
```

Comment:
Side comment: `kubo` supports [custom 404 pages](https://github.com/ipfs/kubo/commit/dfceafdbd3debca60959bdb9ecb28dd4c431c163) already. It does not appear to be in the HTTP gateway spec.
Is this in the scope of this IPIP to provide backward compatibility to this unspecified behavior?

cc @lidel

Reply:
Per a conversation with @lidel, that pretty 4045 functionality isn't officially documented, so we're not including it in the spec.  I'm still falling back to that logic in the code but we plan on removing that code sometime down the road.

Reply:
Ah, and now there's https://github.com/ipfs/ipfs-docs/issues/1155.